### PR TITLE
line: Draw line between element borders

### LINKED ISSUE #234: Drawing well-centered lines between elements
Currently, there is no simple way to draw a "well-centered line" between 2 elements in CeTZ.

I call a line between two elements "well-centered" if it follows the same path as the infinite line that passes through the centers of both elements, but it just touches the edges of the 2 elements instead. I know that this may not be very clear, so this is the kind of things that I want to achieve (this was produced in TikZ):

The solution should also be element- and position-agnostic, so no cheating like for instance assuming that the elements are circles of a given radius, or assuming their position and inferring a given line angle.

Doing this in TikZ is very easy:
```latex
\node[draw, circle] (A) at (0,0) {arbitrary text};
\node[draw, align=left] (B) at (2,3) {arbitrary\\text};
\draw[<->] (A) -- (B);
```

So the solution in CeTZ should be as intuitive as the TikZ solution. Something like:
```typst
content((0,0), [arbitrary text], name: "A", frame: "circle")
content((2,3), [arbitrary\ text], name: "B", frame: "rect")
line("A", "B", mark: (end: ">", start: ">"))
```

Such a feature could be very helpful for a variety of things, such as drawing pretty box-and-arrow diagrams. It could also be useful for the tree CeTZ library (see [this other issue](https://github.com/johannes-wolf/cetz/issues/225)) to replicate the behaviour of the tree TikZ library (the lines between the elements are "well-centered", and not drawn between the `top` and `bottom` anchors of the elements).

See [this Discord thread](https://discordapp.com/channels/1054443721975922748/1154880386942390366) for more info.

### PR DESCRIPTION
**Pull request by johannes-wolf**
Fixes #234

If last and or first argument of `line` is an element name, use that for intersection test.
Saves complete elements in `ctx.nodes`.

### CODE REVIEW

File: src/anchor.typ [hunk 1]
```diff
@@ -109,12 +109,23 @@
       continue
     }
     pts += intersection.line-path(..test-line, drawable)
+  }
+  assert(pts.len() > 0, message: strfmt("{} {} {}", test-line, drawables, angle))
```
Comment (fenjalien) [inline src/anchor.typ:113]: Remove this

File: src/draw/shapes.typ [hunk 2]
```diff
@@ -582,9 +594,72 @@
...
+        vector.as-vec(pt, init: (0,0,0,1))).slice(0, 3)
+    }
+  }
   
   return (ctx => {
+    let first-elem = pts.first()
+    let first-is-elem = coordinate.resolve-system(first-elem) == "element"
+
+    let last-elem = pts.last()
+    let last-is-elem = coordinate.resolve-system(last-elem) == "element"
```
Comment (fenjalien) [inline src/draw/shapes.typ:635]: Use the results of `resolve-system` from line 596 above.

**Review by fenjalien — CHANGES_REQUESTED** [review]

File: tests/intersection/test.typ [hunk 3]
```diff
@@ -69,7 +69,7 @@
     content((0, 0), [Das ist\ ein Text!], frame: "circle", name: "a")
     content((2, 1), [Hallo!], frame: "circle", name: "b")
     // Invisible intersection line
-    line("a", "b", stroke: none)
+    line("a.default", "b.default", stroke: none)
```
[the remark this answers is not in the document]
Comment (johannes-wolf) [reply tests/intersection/test.typ:72]: Because we do not want to calculate intersections here.

File: src/draw/shapes.typ (hunk 2 above)
Comment (johannes-wolf) [reply src/draw/shapes.typ:625]: What do you mean? I am using the result to compare against "element"

File: src/draw/shapes.typ [hunk 4]
```diff
@@ -556,8 +556,20 @@
 /// line((-1.5, 0), (1.5, 0))
 /// line((0, -1.5), (0, 1.5))
 /// line((-1, -1), (-0.5, 0.5), (0.5, 0.5), (1, -1), close: true)
-/// ```) 
-/// 
+/// ```)
+///
+/// If the first or last coordinate given is the name of an element,
+/// the intersection of that element and a line from the first or last
+/// two coordinates given is used as coordinate. This is useful to
```
Comment (fenjalien) [inline src/draw/shapes.typ:563]: What about "If the first or last points are given as the name of an element, that has a "default" anchor, the intersection of that element's border and the line."

**Review by fenjalien — COMMENTED** [review]
I'm having double thoughts about not using border anchors for this. If we do custom elements can define their own border instead relying on something you can't easily configure...

**Comment by johannes-wolf** [comment]
> I'm having double thoughts about not using border anchors for this. If we do custom elements can define their own border instead relying on something you can't easily configure...

I do not get what you mean. Custom elements should be able to override the behavior of how the intersection is determined?

**Pull request merged**